### ISSUE TITLE
User directories (like apache's UserDir)

### ISSUE THREAD

**Issue by ndaniels** [joseluisq/static-web-server#321]
### Search for duplicate feature request

- [X] I already searched, and this feature request or improvement is not a duplicate.

### Feature scope

Completely new feature

### Describe the solution you'd like

The ability to specify (in the config file) a UserDir-like directive, so that a directory such as `public_html` in each user's home directory could be served. Essentially, duplicate this functionality from Apache.

This isn't as common a use case as it once was, but for (for instance) a computer science department webserver, I'd love to be able to move away from Apache and to a Rust-based solution.

### Describe alternatives you've considered

Stick with Apache

### Build target

All targets

**Comment by joseluisq** [comment]
I believe having a [UserDir](https://httpd.apache.org/docs/2.4/mod/mod_userdir.html)-like SWS feature will be useful to entities like computer science departments and similar.

**Comment by lunasophia** [comment]
What would this feature look like? Back in my sysadmin days I only used the most basic features of the UserDir module: given a `/~foo/index.html` URL, serve `/home/foo/public_html/index.html`. Looking at the docs for UserDir reveals many different ways to serve files from user home directories.

Additionally, we shouldn't just serve any user's home directory--the docs for the UserDir module rightly point out that allowing `/~root/` is problematic. It seems like we should start with a default deny with an explicit list of allowed users.

That sysadmin experience means I'm inclined to start looking at this from the sysadmin's perspective, viz. how they configure sws to do this, and then worry about implementation after we've gotten config syntax we like. So with that in mind, here's a first draft config based on the simple case:

```toml
[userdir]
users = ["foo", "bar"]
userdir = "public_html"
```

This would do the same as the simple syntax for the Apache httpd module: search for `/home/{foo,bar}/public_html` and serve files from those directories. This does limit the feature to the two users given. I have a strong preference for explicitly listing the allowed / denied users, unlike the httpd module, which puts everything on one line: `UserDir enabled user1 user2 user3` or `UserDir disabled user4 user5 user6`. I don't see a reason to do this when it's possible to just add another key to the config.

Besides my own simple use case, I'm especially interested in hearing from OP and their CS department usage, which may invalidate some assumptions I've put forward here.

**Comment by ndaniels** [comment]
Very definitely should not serve every user's home directory as you point out. Ideally it would be good to support not just users but groups in the allowed/denied lists. I like the syntax you suggest, though a 'groups' key would be nice (and then use unix groups to configure who gets a public_html dir).

**Comment by lunasophia** [comment]
The challenge I see with determining whether users in a group have their `public_html` served is that group membership is dynamic[^dynamic]. This means that a user can gain or lose group membership (and thus the privilege of having their `public_html` served) after startup. I see two ways to solve this: fetch the group membership at startup or fetch it on every request. I'm leaning more toward doing it at startup, along with username validation.

This raises another question: is it a fatal error to specify an unknown user or group in the config? I can see arguments for both allowing and forbidding this.

[^dynamic]: Okay, usernames are, too, but generally usernames change a lot less than group membership does

**Comment by ndaniels** [comment]
Certainly, group membership *could* be dynamic, even chaotic. But on a well-run system it shouldn't be -- set up a 'webusers' group and include appropriate regular users in it. Certainly you'd add new users over time, and add them to that group. You probably wouldn't routinely remove them from the group unless you disabled their account.

I would suspect that for performance reasons, you'd want to fetch group membership at startup or upon receiving a SIGHUP, so standard procedure when adding a new user would be to kill -HUP the webserver (or restart the daemon). Same if removing a user.

As far as a fatal error, I think that should be configurable with some sort of `--paranoid` flag. At the minimum, a warning should be logged.

**Comment by joseluisq** [comment]
Thanks for the feedback on this. As you discussed, I'm also in favor of being explicit when exposing user home directory content.

The configuration proposal is concise enough in my opinion, which could be extended in the future by adding opt-in directory listing, basic auth, etc per or even allowing multiple `[userdir]` entries per config.

About groups, I vote for fetching the group membership at startup time for performance reasons. But to keep the group membership data synced, we could go with a zero-downtime syncing approach via some mechanism like signaling (SIGNUPs) so SWS could decide whether to serve or not accordingly without restarting. Because the other option should be to restart the server on demand.
Or maybe restarting could be the option in the first stage of the eventual implementation and then enhanced with a zero-downtime syncing approach, just ideas.

Please feel free to continue discussing and suggesting.

**Comment by lunasophia** [comment]
On Fri, Jul 19, 2024 at 03:08:48 -0700, Jose Quintana wrote:
> Thanks for the feedback on this. As you discussed, I'm also in favor
> of being explicit when exposing user home directory content.
> 
> The configuration proposal is concise enough in my opinion, which
> could be extended in the future by adding opt-in directory listing,
> basic auth, etc per or even allowing multiple `[userdir]` entries per
> config.

Agreed. I wanted to stick to the same style of config sws is already
using. I could see an argument for placing `[userdir]` under
`[advanced]` and I'd be fine with that as well.

> About groups, I vote for fetching the group membership at startup time
> for performance reasons. But to keep the group membership data synced,
> we could go with a zero-downtime syncing approach via some mechanism
> like signaling (SIGNUPs) so SWS could decide whether to serve or not
> accordingly without restarting. Because the other option should be to
> restart the server on demand. Or maybe restarting could be the option
> in the first stage of the eventual implementation and then enhanced
> with a zero-downtime syncing approach, just ideas.

Everything here makes sense too.

> Please feel free to continue discussing and suggesting.

Will do! Speaking of first stage implementation, I'd like to start
poking at it this weekend. Once I have something that seems to work I'll
make a draft PR and we can go from there. Sound good?

**Comment by joseluisq** [comment]
> Will do! Speaking of first stage implementation, I'd like to start
poking at it this weekend. Once I have something that seems to work I'll
make a draft PR and we can go from there. Sound good?

Sure, please go ahead and let me know if you need any help.
